aten/src/ATen/core/ATenDispatch.h

```diff
@@ -73,34 +74,108 @@ class CAFFE2_API ATenDispatch {
+      }
+      op_tables_.at(schema).registerOp(backend, reinterpret_cast<void*>(fn));
     }
-    op_tables_.at(schema).registerOp(backend, reinterpret_cast<void*>(fn));
     return *this;
   }
 
   template <class FuncType>
   ATenDispatch& registerVariableOp(const char* schema, FuncType* fn) {
-    std::lock_guard<std::mutex> lock(mutex_);
-    if (op_tables_.find(schema) == op_tables_.end()) {
-      op_tables_.insert(std::make_pair(schema, ATenOpTable(schema)));
+    if (op_was_already_moved_to_c10(schema)) {
+      add_c10_op_registration_(
+        torch::RegisterOperators().op(schema, torch::RegisterOperators::options()
+          .impl_unboxedAutogradKernel(fn)
+          .aliasAnalysis(c10::AliasAnalysisKind::FROM_SCHEMA)
+      ));
+    } else {
+      std::lock_guard<std::mutex> lock(mutex_);
+      if (op_tables_.find(schema) == op_tables_.end()) {
+        op_tables_.insert(std::make_pair(schema, ATenOpTable(schema)));
+      }
+      op_tables_.at(schema).registerVariableOp(reinterpret_cast<void*>(fn));
     }
-    op_tables_.at(schema).registerVariableOp(reinterpret_cast<void*>(fn));
     return *this;
   }
 
   const ATenOpTable* getOpTable(const char* schema) const {
+    TORCH_CHECK(!op_was_already_moved_to_c10(schema), "Tried to get the globalAtenDispatch table for an op that is already moved to the c10 dispatcher");;
     auto iter = op_tables_.find(schema);
     TORCH_CHECK(iter != op_tables_.end(),
         "No functions are registered for schema ", schema);
     return &iter->second;
   }
 
  private:
+  static bool op_was_already_moved_to_c10(const char* schema) {
+    auto op_name = parse_operator_name(schema);
+    return aten_ops_already_moved_to_c10().count(op_name) != 0;
+  }
+
+  static c10::OperatorName parse_operator_name(const char* schema) {
+    // We can't depend on the jit function schema parser here, but parsing
+    // the op name is trivial. Let's just do it by hand.
+    std::string schema_str(schema);
+    size_t name_end_pos = schema_str.find_first_of(".(");
+    TORCH_CHECK(name_end_pos != std::string::npos, "Operator schema must contain a '(' character to start the argument list");
+    size_t overload_name_end_pos = name_end_pos + 1;
+    if (schema_str[name_end_pos] == '.') {
+      overload_name_end_pos = schema_str.find_first_of('(', name_end_pos);
+      TORCH_INTERNAL_ASSERT(schema_str[overload_name_end_pos] == '(');
+    }
+    return c10::OperatorName{
+      schema_str.substr(0, name_end_pos),
+      schema_str.substr(name_end_pos + 1, overload_name_end_pos - name_end_pos - 1)
+    };
+  }
+
   std::unordered_map<std::string, ATenOpTable> op_tables_;
   std::mutex mutex_;
+
+  // Store a torch::RegisterOperators object so it isn't destructed
+  // (which would deregister it)
+  void add_c10_op_registration_(torch::RegisterOperators&& registration) {
+    // This function stores a vector<RegisterOperators> using a singleton pattern.
```

Comment:
The implementation strategy in this file is really weird to me. What I thought would have happened is that ATen native operators which had been moved to c10 would have had c10 registration code generated for them. Instead, what seems to be happening is that you're reusing the ATenDispatch::registerOp calls, and then shoving the c10 registration code inside those function calls. But then you have a problem where they get destructed immediately on exit of the function, so you shoved them in a vector... and then you had the ordering problem which is what this comment is about. But this rube goldberg machine seems totally unnecessary: just codegen c10 registrations top level and you're done! What am I missing?

Reply:
What you're thinking is right. We are codegen'ing c10 registrations, see the diff right previous to this one (https://github.com/pytorch/pytorch/pull/23667). This code here makes sure that backend extensions don't break until we want to break them, see the plan in https://github.com/pytorch/pytorch/issues/24132